### PR TITLE
DBZ-1904 Public facing schema history topic

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.redhat.com/browse/DBZ-1904

### CODE REVIEW

**Comment by gunnarmorling** [comment]
@jpechane, can you add the documentation update (new option, probably example records)?

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java [hunk 1]
```diff
@@ -395,6 +397,10 @@ public boolean getSanitizeFieldNames() {
         }
     }
 
+    public boolean includeSchemaChangeRecords() {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:400]: Can you add a comment? I know it wasn't there before, but it's a good practice for those framework-level classes, and potential future implementors might wonder about its purpose.

Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:400]: Maybe we also should rename it, e.g. `isSchemaChangesHistoryEnabled()` or similar?

File: debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java [hunk 2]
```diff
@@ -241,6 +263,19 @@ public void dispatchSchemaChangeEvent(T dataCollectionId, SchemaChangeEventEmitt
         schemaChangeEventEmitter.emitSchemaChangeEvent(new SchemaChangeEventReceiver());
     }
 
+    public void dispatchSchemaChangeEvent(List<T> dataCollectionIds, SchemaChangeEventEmitter schemaChangeEventEmitter) throws InterruptedException {
+        final boolean anyNonfilteredEvent = dataCollectionIds.stream()
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java:267]: Let's avoid stream ops in this inner loop method.

File: debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java [hunk 3]
```diff
@@ -267,6 +302,10 @@ public void dispatchTransactionMessage(SourceRecord record) throws InterruptedEx
         queue.enqueue(new DataChangeEvent(record));
     }
 
+    public void dispatchSchemaChangeMessage(SourceRecord record) throws InterruptedException {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java:305]: How is it different from the method above?

File: debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java [hunk 4]
```diff
@@ -275,6 +275,16 @@ public static DecimalHandlingMode parse(String value, String defaultValue) {
...
             .withDescription("The maximum number of millis to wait for table locks at the beginning of a snapshot. If locks cannot be acquired in this " +
                     "time frame, the snapshot will be aborted. Defaults to 10 seconds");
+    public static final Field INCLUDE_SCHEMA_CHANGES = Field.create("include.schema.changes")
+            .withDisplayName("Include database schema changes")
+            .withType(Type.BOOLEAN)
+            .withWidth(Width.SHORT)
+            .withImportance(Importance.MEDIUM)
+            .withDescription("Whether the connector should publish changes in the database schema to a Kafka topic with "
+                    + "the same name as the database server ID. Each schema change will be recorded using a key that "
+                    + "contains the database name and whose value includes the DDL statement(s)."
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java:285]: DDL statement is not always correct, right?

File: debezium-core/src/main/java/io/debezium/relational/history/TableChanges.java [hunk 5]
```diff
@@ -199,98 +139,26 @@ else if (!table.equals(other.table)) {
...
-                    .forEach(editor::addColumn);
+    /**
+     * The interface that defines conversion of {@code TableChanges} into a serialized format for
+     * persistent storage or delivering as a message.
+     *
+     * @author Jiri Pechanec
+     *
+     * @param <T> target type
+     */
+    public static interface TableChangesSerializer<T> {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/history/TableChanges.java:152]: Can you clarify why there are two implementations?

**Comment by gunnarmorling** [comment]
@jpechane, are you still planning to address my comments above?

File: debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java (hunk 1 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java]: Done

File: debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java (hunk 2 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java]: Done

File: debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java (hunk 3 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java:310]: It is the same but the trigger to generate the message is different. We could merge it into a single method but I like keeping it separated - it is easy to add further logging/metrics etc.

File: debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java (hunk 4 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java]: Fixed

File: debezium-core/src/main/java/io/debezium/relational/history/TableChanges.java (hunk 5 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/relational/history/TableChanges.java:152]: one is for internal representation of schema history - directly written into the JSON.
The other is public facing structure for external consuming - can be JSON/Avro according to the users needs and also public and private format will not affect each other.

**Comment by gunnarmorling** [comment]
@jpechane, can you add a commit for the doc update? Thanks!

File: debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java [hunk 6]
```diff
@@ -275,6 +275,16 @@ public static DecimalHandlingMode parse(String value, String defaultValue) {
             .withDefault(DEFAULT_SNAPSHOT_LOCK_TIMEOUT_MILLIS)
             .withDescription("The maximum number of millis to wait for table locks at the beginning of a snapshot. If locks cannot be acquired in this " +
                     "time frame, the snapshot will be aborted. Defaults to 10 seconds");
+    public static final Field INCLUDE_SCHEMA_CHANGES = Field.create("include.schema.changes")
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java:278]: @jpechane Looks like it's not added to any config group? Before doing so, probably a good idea to rebase this one, so you can do it the new way after the recent refactoring.

**Comment by gunnarmorling** [comment]
@jpechane, I'm having some issues with testing this. Using the tutorial example, I dropped a column, but I see no schema change event. Also, I see the original entries, but then there are some ALTER statements (e.g. for the order table), where are those coming from? Specifically, can we add the original SQL DDL statement to events, too?

File: debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java [hunk 7]
```diff
@@ -241,6 +263,24 @@ public void dispatchSchemaChangeEvent(T dataCollectionId, SchemaChangeEventEmitt
         schemaChangeEventEmitter.emitSchemaChangeEvent(new SchemaChangeEventReceiver());
     }
 
+    public void dispatchSchemaChangeEvent(List<T> dataCollectionIds, SchemaChangeEventEmitter schemaChangeEventEmitter) throws InterruptedException {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java:266]: I don't find any callers to this method.

File: debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java (hunk 3 above)
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java:310]: Ok, so this should be private and also named `enqueueSchemaChangeMessage()`. Similar for `dispatchTransactionMessage()` right above. Those methods act on a different level than the dispatchXyz ones.

File: debezium-core/src/main/java/io/debezium/relational/history/ConnectTableChangeSerializer.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        struct.put(OPTIONAL_KEY, column.isOptional());
+        struct.put(AUTO_INCREMENTED_KEY, column.isAutoIncremented());
+        struct.put(GENERATED_KEY, column.isGenerated());
+
+        return struct;
+    }
+
+    @Override
+    public TableChanges deserialize(List<Struct> data, boolean useCatalogBeforeSchema) {
+        throw new UnsupportedOperationException("Deserialization fro Connect Struct is not supported");
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/history/ConnectTableChangeSerializer.java:143]:
```suggestion
        throw new UnsupportedOperationException("Deserialization from Connect Struct is not supported");
```

File: debezium-core/src/main/java/io/debezium/relational/history/TableChanges.java (hunk 5 above)
Comment (gunnarmorling) [reply debezium-core/src/main/java/io/debezium/relational/history/TableChanges.java:152]: Ok, thanks for clarifying.

File: debezium-core/src/main/java/io/debezium/schema/SchemaChangeEvent.java [hunk 9]
```diff
@@ -26,24 +29,36 @@
...
         this.ddl = ddl;
         this.tables = Objects.requireNonNull(tables, "tables must not be null");
         this.type = Objects.requireNonNull(type, "type must not be null");
         this.isFromSnapshot = isFromSnapshot;
+        switch (type) {
+            case CREATE:
+                tables.forEach(tableChanges::create);
+            case ALTER:
+                tables.forEach(tableChanges::alter);
+            case DROP:
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/schema/SchemaChangeEvent.java:60]: This looks suspicous? If nothing is to be done in this case indeed, a comment is warranted.

File: debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SchemaHistoryTopicIT.java [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+import io.debezium.config.Configuration;
+import io.debezium.connector.sqlserver.SqlServerConnectorConfig.SnapshotMode;
+import io.debezium.connector.sqlserver.util.TestHelper;
+import io.debezium.doc.FixFor;
+import io.debezium.embedded.AbstractConnectorTest;
+import io.debezium.relational.RelationalDatabaseConnectorConfig;
+import io.debezium.util.Testing;
+
+/**
+ * Integration test for the Debezium SQL Server connector.
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SchemaHistoryTopicIT.java:29]: Might be a bit more specific.

File: debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java (hunk 7 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/pipeline/EventDispatcher.java:266]: @gunnarmorling This one is oging to be used when MySQL is migrated to the new framework.

**Comment by jpechane** [comment]
@gunnarmorling Code should be finalized now, I am working on the docs

> Why don't I see events in case of dropping a column? Same for ADD COLUMN btw., i.e. I don't see any change event during streaming.
The schema migration procedure for SQL Server needs to be followed.

> Where do the ALTER statements I see come from?
That was embrassing bug - Switch in `io.debezium.schema.SchemaChangeEvent.SchemaChangeEvent(Map<String, ?>, Map<String, ?>, Struct, String, String, String, Set<Table>, SchemaChangeEventType, boolean)` lacked breaks

> Why can one record have multiple events? Wouldn't be better to have one record per event?
That's taken over from MySQL.

Docs should be added; we should describe this format as incubating, I think

**Comment by gunnarmorling** [comment]
Rebased and applied. Thanks, @jpechane!

>> Why can one record have multiple events? Wouldn't be better to have one record per event?
> That's taken over from MySQL.

I'm still pondering about this one; but it's described as incubating, so we can revisit this in the course of the MySQL work.

**Pull request closed without merging**